pkg/controller/bastion/actuator_reconcile.go

```
@@ -0,0 +1,437 @@
+	}
+
+	// check if the instance already exists and has an IP
+	endpoints, err := getInstanceEndpoints(&instances[0], opt)
+	if err != nil {
+		return err
+	}
+
+	if !endpoints.Ready() {
+		return &ctrlerror.RequeueAfterError{
+			// requeue rather soon, so that the user (most likely gardenctl eventually)
+			// doesn't have to wait too long for the public endpoint to become available
+			RequeueAfter: 5 * time.Second,
+			Cause:        errors.New("bastion instance has no public/private endpoints yet"),
+		}
+	}
+
+	// once a public endpoint is available, publish the endpoint on the
+	// Bastion resource to notify upstream about the ready instance
+	patch := client.MergeFrom(bastion.DeepCopy())
+	bastion.Status.Ingress = endpoints.public
+	return a.client.Status().Patch(ctx, bastion, patch)
+}
+
+func ensurePublicIPAddress(opt *Options, openstackClientFactory openstackclient.Factory) (*floatingips.FloatingIP, error) {
+	fips, err := getFipByName(openstackClientFactory, opt.BastionInstanceName)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(fips) != 0 && fips[0].Status == "ACTIVE" {
+		return &fips[0], nil
+	}
+
+	logger.Info("creating new bastion Public IP")
+
+	externalFipInfo, err := getExternalNetworkInfoByName(openstackClientFactory, opt.FloatingPoolName)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(externalFipInfo) == 0 {
+		return nil, errors.New("externalFipInfo must not be empty")
+	}
+
+	createOpts := floatingips.CreateOpts{
+		FloatingNetworkID: externalFipInfo[0].ID,
+		Description:       opt.BastionInstanceName,
+	}
+
+	fip, err := createFloatingIP(openstackClientFactory, createOpts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get (create) public ip address: %w", err)
+	}
+
+	return fip, nil
+}
+
+func ensureComputeInstance(logger logr.Logger, openstackClientFactory openstackclient.Factory, opt *Options) (*servers.Server, error) {
+	instances, err := getBastionInstance(openstackClientFactory, opt.BastionInstanceName)
```

err is not handled properly. Only Ignore not found case is covered.